File: UEMPS.cpp
#include "UEMPS.h"

int main(int argc, char **argv) {
    int counter = 0;

    zmq::context_t ctx;
    msgpack::zone z;
    zmq::socket_t sock(ctx, zmq::socket_type::pub);
    sock.bind("tcp://127.0.0.1:6666");

    std::this_thread::sleep_for(std::chrono::milliseconds(1000));

    while (true) {
        std::this_thread::sleep_for(std::chrono::milliseconds(100));

        MSG msg;

        if (counter % 2 == 0) {
            STATE_PLAYER state_player = STATE_PLAYER{9, "warrior", 26};
            msg = MSG{MSG_ID::STATE_PLAYER, msgpack::object(state_player, z)};
            std::cout << " [server] " << counter <<  " warrior \n";
        } else {
            STATE_NPC state_npc = STATE_NPC{200, "salesman", 98, 100};
            msg = MSG{MSG_ID::STATE_NPC, msgpack::object(state_npc, z)};
            std::cout << " [server] " << counter <<  " salesman \n";
        }

        std::stringstream buffer;
        msgpack::pack(buffer, msg);
        const std::string payload = buffer.str();

        sock.send(zmq::buffer(payload), zmq::send_flags::dontwait);

        counter++;
    }
}

File: UEMPS.py
import time
import threading
import zmq
import msgpack
import UEMP

# global variables
players = {}


def publisher():
    """This function is used to create a publisher that sends messages to the subscribers"""

    # create ZeroMQ server
    context = zmq.Context()
    socket = context.socket(zmq.PUB)
    socket.bind("tcp://127.0.0.1:6666")

    # infinite loop
    while True:

        # send all the players
        for player in players.keys():
            # create a message object
            message = UEMP.MSG(UEMP.MSG_ID.STATE_PLAYER, players[player])

            # print the message
            # print("Server:", round(time.time(), 3), "send", message)

            # serialize the object with its method
            packed = message.pack()
            serialized = msgpack.packb(packed)

            # send the message with no wait
            socket.send(serialized, zmq.DONTWAIT)

        # wait for some time
        time.sleep(0.1)


def puller():
    """This function is used to create a puller that receives messages from the pushers"""

    # create ZeroMQ server
    context = zmq.Context()
    socket = context.socket(zmq.PULL)
    socket.bind("tcp://127.0.0.1:6667")

    # infinite loop
    while True:
        # receive the message
        serialized = socket.recv()

        # deserialize the object
        packed = msgpack.unpackb(serialized)
        message = UEMP.MSG()
        message.unpack(packed)
        player = message.msg

        # print the message
        print("Server:", round(time.time(), 3), "recv", player)

        # find the player in the list or add it if it is not there
        players[player.id] = player


def main():
    """This function is used to create the publisher and the puller"""

    # create the publisher
    publisher_thread = threading.Thread(target=publisher)
    publisher_thread.start()

    # create the puller
    puller_thread = threading.Thread(target=puller)
    puller_thread.start()

    # wait for the threads to finish
    publisher_thread.join()
    puller_thread.join()


if __name__ == "__main__":
    main()
